---
title: "PBMC_10k_v3_Seurat_R Notebook"
 
1) Load UMI count matrix obtained from the CellRanger Pipeline,10x Genomics 
```{r}
library(dplyr)      #Package for Data manipulation              
library(Seurat)     #Package for Single Cell Data Analysis to find cell clusters          
pbmc.data <- Read10X(data.dir = "/Users/swatz_bachoo/Downloads/Sanofi/Seurat/filtered_feature_bc_matrix") #10X PBMCS from Healthy Donor
```

2) Exploring the UMI count sparse matrix in pbmc.data, where,
    Rows = Genes/Features.
    Columns = Cells.
```{r}
cat("Number of rows/genes =", nrow(pbmc.data),"\n")
cat("\nNumber of columns/cells =", ncol(pbmc.data), "\n")
cat ("Sample of pbmc.data\n")
pbmc.data[1:10,30:50] #UMI count of first 10 features from cells 30 to 50
```


3) Advantages of using a sparse matrix
```{r}
#dgCMatrix class is a class of sparse numeric matrices in the compressed, sparse, column-oriented format
dense.size <- object.size(as.matrix(pbmc.data))
sparse.size <- object.size(pbmc.data)
memorySavings <- dense.size/sparse.size
cat("Memory Savings (Fold change) when using a sparse matrix than a dense matrix =",memorySavings, "\n")
```


4) Create Seurat Object "pbmc" with the raw data.
   Selecting cells which have atleast 200 unique genes and genes which are expressed in atleast 3 cells.
```{r}
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc10k", min.cells = 3, min.features = 200)
pbmc
```


5) Pre-Processing: Number of total molecules and unique genes are stored in meta.data while creating seurat object. Additional QC metrices can be added as columns here.
nCount_RNA: total number of molecules detected within a cell.
nFeature_RNA : number of unique genes detected in each cell.
```{r}
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
head(pbmc@meta.data, 5) #New QC metric column "percent.mt" is added now
```
   
   
6) Visualization of QC Metrics:
   (a) FeatureScatter: Used to visualize feature-feature relationships. Total number of molecules detected within a cell correlates strongly with unique genes.
```{r}
library(ggplot2)
plot1 <- FeatureScatter(object = pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt", pt.size = 0.05) + theme(axis.text = element_text(size = 8))
plot2 <- FeatureScatter(object = pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", pt.size = 0.05) + theme(axis.text = element_text(size = 8))
CombinePlots(plots = list(plot1, plot2))
  #r: Strength and Directionality of a linear relationship
  #nCount_RNA & percent.mt : Weak Negative Linear Relationship
  #nFeature_RNA & percent.mt : Strong Positive Linear Relationship

```
 
 
   (b) Violin PLots: Black dots represent the values for individual cells. Wider sections represent a higher probability of a given value; the skinnier sections represent a lower probability.
```{r}
vlnplot1 <- VlnPlot(object=pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0.03)
```


7) QC: 
    Unique gene count/cell should neither be too high (cell doublets/multiplets) nor too low (empty droplets/low quality cells) - cutoff set at not over than 2,500 nor less than 200.
    Percentage of reads mapping to mt genome should be less than 5%.
```{r}
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
require(gridExtra)
vlnplot2 <- VlnPlot(object=pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0.1)
grid.arrange(vlnplot1, vlnplot2, nrow = 2)
```


8) Data Normalization:
    (i) Feature expression/cell divided by total expression (To compare gene expression across cells).
    (ii) Scale factor of 10000 is multiplied to avoid negative values after Ln Transformation. Now, sum of all transcript counts across all genes in each cell will be 10,000.
    (iii) LogNormaliztion used is a natural log transform (Inclusion of wide ranges, account for skewness, more symmetric distribution of data, intutive sense of fold changes).
```{r}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
#pbmc[["RNA"]]@data
```

9) Feature Selection: 
   Identifying features with high cell-to-cell variation for biological insights.
   FindVariableFeature(): Identifies features that are outliers on a mean variability plot.
                          (a) Comparison across features.
                          (b) Computes variance of standardized values across all cells in a way similar to a z-transformation based on expected varaince from the expected means of gene expression (Local polynomial regression). 
                          (c) Used directly to rank the features. 
```{r}
#nFeatures: Number of features to select as top variable features
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000) 
top15 <- head(VariableFeatures(pbmc), 15)
plot3 <- VariableFeaturePlot(pbmc)
plot4 <- LabelPoints(plot = plot3, points = top15, repel = TRUE, xnudge = 0, ynudge = 0)
plot4
```
   
10) Linear Tansformation/Scaling across cells [(X-Mean)/Variance]: Done so that highly-expressed genes do not dominate in the analysis results.
    Standard Normal Distribution:
    Mean expression of each gene across cells = 0;
    Variance in expression of each gene across cells = 1. 
```{r}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes) #Scaling all features
#pbmc <- ScaleData(pbmc) Scaling only the highly variable features
#pbmc[["RNA"]]@scale.data
```
 
11) Removing known unwanted variation: Focussing only on genomic DNA variation
```{r}
pbmc <- ScaleData(pbmc, vars.to.regress = "percent.mt")
```

12) Linear Dimensional Reduction: Principal Component Analysis (PCA).
      (a) Supervised Method for determining relevant sources of heterogeneity.
      (b) Each PC essentially representing a metafeature that combines information across a correlated feature set. 
      (c) The top principal components therefore represent a robust compression of the dataset.
```{r}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
#print(pbmc[["pca"]], dims = 1:5, nfeatures = 5) #Top 5 
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca") #Visualize top genes associated with reduction components
DimPlot(pbmc, reduction = "pca") #Orthogonal Variation
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE) 
#Visualization of heterogeneity. 
#Cells and genes are sorted by their principal component scores
#Can see Smaller sub populations with strong separation
#Negatuve and Positive Markers
#DimHeatmap(pbmc, dims = 1:2, cells = 500, balanced = TRUE)
```

13) Determining Dimensionality/Significant PCs: 
    JackStrawPlot: QQ plot to identify significant PCs.If quantiles of PCs came from the same distribution, we should see the points forming a line that’s roughly straight. Distribution of p-values for the genes in the significant PCs lie parallelly close to the uniform distribution (null distribution of p-value feature scores from repeatedly permutated data). 
    ElbowPlot: An ‘elbow’ around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs
```{r}
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
JackStrawPlot(pbmc, dims = 1:15)
ElbowPlot(pbmc) 
```

14) Clustering of Cells:
    K-nearest neighbor (KNN) graph approach to find neighbours, i.e, cells with similar expression patterns.
    Expect Grouping/Clustering of similar cells.
```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:10, reduction = "pca") #Dimensions of reduction to use as input
pbmc <- FindClusters(pbmc, resolution = 2.4) 
#Higher resolution, more communities found. Howvever, overlaps also increase.
head(Idents(pbmc), 10) #Idents() used to find cluster IDs of the first five cells
```
   
15) Non-Linear Dimensional Reduction (tSNE/UMAP): To place similar cells together in low-dimensional space
```{r}
library(reticulate)
library(umap)
pbmc <- RunUMAP(object = pbmc, dims = 1:10)
DimPlot(pbmc, reduction = "umap", label = TRUE, repel = TRUE, xnudge = 0, ynudge = 0) #Same as UMAPPlot()
```

16) Finding differentially expressed features/Cluster Biomarkers:
    (a) min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells
    (b) logfc.threshold argument requires a feature to be differentially expressed (on average) by some amount between the two groups
```{r}
library(magrittr)
#Identifies positive and negative markers of a single cluster (ident.1), compared to all other cells
cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, min.pct = 0.25)
head(cluster1.markers, n = 5)
#Identifies positive and negative markers of a single cluster compared to cells from other specified clusters
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(6, 8), min.pct = 0.25)
head(cluster5.markers, n = 5)
#Identifies positive markers for all clusters
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = log(2))
pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
```

17) Assigning cell identity to clusters: Based on available literature in this case
```{r}
new.cluster.ids <- c("Platelets","Memory CD4+ T","NK", "CD14+ Mono", "Naive CD4+ T", "CD8+ T", "DC", "B", "FCGR3A+ Mono")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 1.5, repel = TRUE, xnudge = 0, ynudge = 0) + NoLegend()
```

18) Additional Visualizations for Marker Expression:
```{r}
#Violin Plots: Expression probability distributions across clusters
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
#Violin Plots of raw counts
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)
#Visualizes feature expression on a tSNE or PCA plot
FeaturePlot(pbmc, features = c("PPBP", "GNLY", "FCN1", "DNAH12"))
#Expression heatmap for given cells and features for top 10 markers
top10 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(pbmc, features = top10$gene,label = TRUE, size = 2.5, angle = 25) + NoLegend() 
```

